tools/scsisnoop.py

```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/python
+# SPDX-License-Identifier: <SPDX License Expression>
+# @lint-avoid-python-3-compatibility-imports
+#
+# scsisnoop  Trace SCSI layer I/O and print details including issuing PID.
+#           For Linux, uses BCC, eBPF.
+#
+# This uses in-kernel eBPF maps to cache process details (PID and comm) by I/O
+# request, as well as a starting timestamp for calculating I/O latency.
+#
+# Copyright (c) 10-Oct-2022, Samsung Electronics.  All rights reserved.
+# This source code is licensed under the Apache License, Version 2.0
+#
+# Written by Weibang Liu<weibang6.liu@samsung.com>.
+
+from __future__ import print_function
+from bcc import BPF
+import argparse
+
+# arguments
+examples = """examples:
+    ./scsisnoop           # trace all SCSI layer's I/O
+"""
+parser = argparse.ArgumentParser(
+    description="Trace SCSI layer I/O",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("--ebpf", action="store_true",
+                    help=argparse.SUPPRESS)
+args = parser.parse_args()
+debug = 0
+
+# define BPF program
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <linux/blkdev.h>
+#include <scsi/scsi_cmnd.h>
+
+// for saving the timestamp and __data_len of each request
+struct start_req_t {
+    u64 ts;
+    u64 data_len;
+};
+
+struct val_t {
+    u64 ts;
+    u32 pid;
+    char name[TASK_COMM_LEN];
+};
+
+struct data_t {
+    u32 pid;
+    u64 rwflag;
+    u64 delta;
+    u64 qdelta;
+    u64 sector;
+    u64 len;
+    u64 ts;
+    char disk_name[DISK_NAME_LEN];
+    char name[TASK_COMM_LEN];
+};
+
+BPF_HASH(start, struct request *, struct start_req_t);
+BPF_HASH(infobyreq, struct request *, struct val_t);
+BPF_PERF_OUTPUT(events);
+
+// cache PID and comm by-req
+int trace_pid_start(struct pt_regs *ctx, struct scsi_cmnd *cmd)
+{
+    struct val_t val = {};
+    u64 ts;
+    struct request *req = cmd->request;
+
+    if (bpf_get_current_comm(&val.name, sizeof(val.name)) == 0) {
+        val.pid = bpf_get_current_pid_tgid() >> 32;
+
+        infobyreq.update(&req, &val);
+    }
+    return 0;
+}
+
+// time SCSI I/O
+int trace_req_start(struct pt_regs *ctx, struct scsi_cmnd *cmd)
+{
+    struct request *req = cmd->request;
+    //bpf_trace_printk("[Andy]trace_start flag: %x\\n", req->cmd_flags);
+
+    struct start_req_t start_req = {
+        .ts = bpf_ktime_get_ns(),
+        .data_len = req->__data_len
+    };
+    start.update(&req, &start_req);
+    return 0;
+}
+
+// output
+int trace_req_completion(struct pt_regs *ctx, struct scsi_cmnd *cmd)
+{
+    struct start_req_t *startp;
+    struct val_t *valp;
+    struct data_t data = {};
+    u64 ts;
+    struct request *req = cmd->request;
+
+    //bpf_trace_printk("[Andy]trace_end flag: %x\\n", req->cmd_flags);
```

Comment:
?

Reply:
OK, I will check again and remove useless annotate.